lib/driver.js

```diff
@@ -223,13 +228,27 @@ class XCUITestDriver extends BaseDriver {
     return status;
   }
 
+  mergeCliArgsToOpts () {
+    for (const arg of Object.keys(ARGS_CONSTRAINTS)) {
+      if (_.has(this.cliArgs, arg)) {
+        log.info(`Using ${arg} value of '${this.cliArgs[arg]}' from CLI`);
+        if (_.has(this.opts, arg)) {
+          log.info(`(This overwrites value of '${this.opts[arg]}' sent in via caps)`);
+        }
+        this.opts[arg] = this.cliArgs[arg];
+      }
+    }
+  }
+
   async createSession (...args) {
     this.lifecycleData = {}; // this is used for keeping track of the state we start so when we delete the session we can put things back
     try {
       // TODO add validation on caps
       let [sessionId, caps] = await super.createSession(...args);
       this.opts.sessionId = sessionId;
 
+      this.mergeCliArgsToOpts();
```

Comment:
what if some options should be handled in base driver do we deliberately skip such?
also, what to do if the provided value should be verified in combination with other options, not as a single? `validateDesiredCaps` is invoked in the super class.

Reply:
> what if some options should be handled in base driver do we deliberately skip such?

I'm not sure what you mean. This is being called after `super.createSession` so basedriver is not going to touch `this.opts` anymore.

> what to do if the provided value should be verified in combination with other options, not as a single?

I'm not sure I understand. Can you give an example? I guess we can always put some more complex logic here for specific cases if this driver needs it.

Reply:
For example:
1. desired capabilities are 

```
{a: 1}
```
2. Command line args are:
```
{b: 3, c: 4}
```
3. Args validation did pass for the given caps
4. Now we merge command line args to caps and become caps like
```
{a: 1, b: 3, c: 4}
```
which would file being provided initially, because this combination is not allowed, but since `validateDesiredCaps` has already been called this will be accepted and corrupt the flow.

Reply:
we have capability validation that depends on other capabilities?

Reply:
if we do, i suppose we could always call `this.validateDesiredCaps(this.opts)` immediately after merging.

Reply:
you mean to perform the validation again if any of actual options have been modified/changed? Yep, this could be done, although there is still a possibility to erroneously fail in such fail: let say a customer provides an erronous caps combination and the server admin provides args to override these caps. Then the check in base driver would fail while it is expected to pass, because CLI args have priority over session caps

Reply:
IMO this failure scenario is acceptable. If a customer gets feedback that their capabilities are non-sensical, that's OK, even if overrides on the part of the sysadmin would make them sensical. The customer can fix their capabilities to something that makes sense, or follow the vendor's documentation to leave out those capabilities entirely. IMO vendors should have a layer on their load balancer, before Appium comes into the picture at all, that gives customers errors if they use capabilities the vendor doesn't support or will always override.

The alternative would be to do all the merging and validation in basedriver, but then that forces drivers into this scheme where args always show up on `this.opts`. What if driver implementers don't want that? The case where capabilities and CLI args overlap is not universal.

Reply:
I'm ok with the above scenario. It would also make sense to return true from mergeCliArgsToOpts if any caps were actually merged, so that we need to do another verification. Otherwise we could skip it as it has been already done in the base driver

Reply:
OK, how is this most recent version then?